src/basic/devnum-util.c

```diff
@@ -57,6 +58,42 @@ int device_path_make_major_minor(mode_t mode, dev_t devnum, char **ret) {
         return 0;
 }
 
+int device_path_make_major_minor2(mode_t mode, dev_t devnum, char **ret) {
+        const char *t;
+        _cleanup_free_ char *syspath = NULL;
+        _cleanup_free_ char *link = NULL;
+        _cleanup_free_ char *fname = NULL;
+
+        /* Generates the /dev/... path given a dev_t */
+
+        if (S_ISCHR(mode))
+                t = "char";
+        else if (S_ISBLK(mode))
+                t = "block";
+        else
+                return -ENODEV;
+
+        if (asprintf(&syspath, "/sys/dev/%s/" DEVNUM_FORMAT_STR, t, DEVNUM_FORMAT_VAL(devnum)) < 0)
+                return -ENOMEM;
+
+        int r = readlink_malloc(syspath, &link);
+        if (r < 0)
+                return r;
+
+        r = path_extract_filename(link, &fname);
+        if (r < 0)
+                return r;
+
+        if (asprintf(ret, "/dev/%s", fname) < 0)
```

Comment:
we concatenating strings use strjoin(), not asprintf(). In fact, in this case even use path_join(), since you are joining paths here.

Reply:
also, as per coding style: our functions should initialize all return parameters on success, but none on failure. Here you are initializing ret already, even though the stat() below can still fail. If it does then you will have initialized ret even though the function overall returned an error. hence, please write the path to a local variable, and move it into `*ret` only after stat() succeeded.